File: 백준/Gold/1918. 후위 표기식/후위 표기식.cc
#include <iostream>
#include <vector>
#define rep(i, a, b) for(auto i = a; i < b; ++i)
#define REP(i, a, b) for(auto i = a; i <= b; ++i)
#define FAST ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL)

using namespace std;

enum Operator {
    Addition, Substraction, Multiplication, Division, Open, Close
};

const vector<int> priority = {0, 0, 1, 1, 2, 2};
const vector<char> operatorString = {'+', '-', '*', '/', '(', ')'};

int getOperatorIndex(char c) {
    switch (c) {
        case '+':
            return Addition;
        case '-':
            return Substraction;
        case '*':
            return Multiplication;
        case '/':
            return Division;
        case '(':
            return Open;
        case ')':
            return Close;
        default:
            return -1;
    }
}

int main() {
    FAST;
    string infix;
    vector<Operator> stack;
    string res = "";

    cin >> infix;

    rep(i, 0, infix.size()) {
        int curOp = getOperatorIndex(infix[i]);

        // 피연산자 그대로 추가
        if (curOp == -1) {
            res += infix[i];
            continue;
        }

        // 닫는 연산자라면
        if (curOp == Close) {
            while (true) {
                Operator op = stack.back();
                stack.pop_back();

                if (op == Open) { break; }

                res += operatorString[op];
            }
            continue;
        }

        // 현재 연산자보다 우선 순위가 크거나 같은걸 뺀다.
        while (!stack.empty()) {
            Operator preOp = stack.back();
            
            if (preOp == Open) { break; }
            if (priority[curOp] > priority[preOp]) { break; }
            stack.pop_back();

            res += operatorString[preOp];
        }

        stack.push_back((Operator)curOp);
    }

    // 스택에 남아있는 연산자들을 결과에 추가
    while (!stack.empty()) {
        Operator op = stack.back();
        stack.pop_back();

        res += operatorString[op];
    }

    cout << res;
}

File: 백준/Gold/1918. 후위 표기식/후위 표기식.swift
import Foundation

// 중위 표기식 -> 후위 표기식으로 바꾸기
// 피 연산자는 그대로 출력
// 연산자
//  스택이 비어있으면 넣는다.
//  본인보다 우선 순위가 같거나 높은 연산자가 없을 때까지 스택의 연산자를 꺼낸다. 여는 괄호는 닫는 괄호가 나오지 않으면 꺼내지 않음!!
//  본인을 넣는다.
enum Operator: String, Comparable {
    case addtion = "+"
    case subtraction = "-"
    case multiplication = "*"
    case division = "/"
    case open = "("
    case close = ")"

    var priority: Int {
        switch self {
        case .addtion, .subtraction:
            return 0
        case .multiplication, .division:
            return 1
        case .open, .close:
            return 2
        }
    }

    static func <(_ lhs: Operator, _ rhs: Operator) -> Bool {
        return lhs.priority < rhs.priority
    }
}

let infixExpression = readLine()!.map { String($0) }
var operatorStack: [Operator] = []
var res = ""

for element in infixExpression {
    if let op = Operator(rawValue: element) {
        // 닫는 괄호일 경우
        if op == .close {
            // 여는 괄호가 나올 때까지 스택에서 요소를 뺀다.
            while true {
                let lastOperator = operatorStack.removeLast()
                if lastOperator == .open { break }
                res += lastOperator.rawValue
            }
            continue
        }


        // 자기와 우선순위가 같거나 같은 것들을 스택에서 뺀다.
        while
            let lastOperator = operatorStack.last,
            op <= lastOperator
        {
            guard lastOperator != .open else { break } // 여는 괄호는 빼지 않음
            res += operatorStack.removeLast().rawValue
        }

        operatorStack.append(op)
    } else {
        res += element
    }
}

while !operatorStack.isEmpty {
    res += operatorStack.removeLast().rawValue
}

print(res)
